### ISSUE TITLE
Move accountant towards atomic balances

### ISSUE THREAD

**Issue by garious** [solana-labs/solana#139]
Below is some awkward code that requires the accountant balance to be locked for multiple steps. If instead we added a new `forget_signature_with_last_id()` before returning `InsufficientFunds`, then we can reserve the signature first (before checking the balance), which opens the door for replacing the `bal` RwLock with an Atomic.

```rust
        if *bal < tr.data.tokens {
            return Err(AccountingError::InsufficientFunds);
        }

        if !self.reserve_signature_with_last_id(&tr.sig, &tr.data.last_id) {
            return Err(AccountingError::InvalidTransferSignature);
        }

        *bal -= tr.data.tokens;
```

**Comment by rlkelly** [comment]
Taking a look at this one today

**Comment by garious** [comment]
FYI https://github.com/garious/solana/commit/ecabce57cdf08154aa23e47547b833824ebb6911

**Comment by garious** [comment]
Fixed in fd5ef94b5a8a944827890a08c26cae8dc1635d5a
